sdr_airspy.c

```diff
@@ -0,0 +1,467 @@
+//
+//    *  Redistributions in binary form must reproduce the above copyright
+//       notice, this list of conditions and the following disclaimer in the
+//       documentation and/or other materials provided with the distribution.
+//
+//   THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+//   "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+//   LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+//   A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+//   HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+//   SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+//   LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+//   DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+//   THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+//   (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+//   OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+#include "dump1090.h"
+#include "sdr_airspy.h"
+
+#include <airspy.h>
+#include <soxr.h>
+#include <inttypes.h>
+
+static struct {
+    struct airspy_device *dev;
+    uint64_t serial_number;
+    bool enable_lna_agc;
+    bool enable_mixer_agc;
+    bool enable_linearity;
+    bool enable_sensitivity;
+    bool enable_airspy_biast;
+    uint8_t linearity_gain;
+    uint8_t sensitivity_gain;
+    uint8_t mixer_gain;
+    uint8_t lna_gain;
+    uint8_t vga_gain;
+    soxr_t resampler;
+    char *airspy_bytes;
+    char *airspy_scratch;
+    iq_convert_fn converter;
+    struct converter_state *converter_state;
+} AIRSPY;
+
+//
+// =============================== AIRSPY handling ==========================
+//
+
+void airspyInitConfig()
+{
+    AIRSPY.dev = NULL;
+    AIRSPY.serial_number = 0;
+    AIRSPY.enable_lna_agc = false;
+    AIRSPY.enable_mixer_agc = false;
+    AIRSPY.enable_linearity = false;
+    AIRSPY.enable_sensitivity = false;
+    AIRSPY.enable_airspy_biast = false;
+    AIRSPY.linearity_gain = 0;
+    AIRSPY.sensitivity_gain = 0; 
+    AIRSPY.mixer_gain = 8;
+    AIRSPY.lna_gain = 13;
+    AIRSPY.vga_gain = 5;
+    AIRSPY.resampler = NULL; 
+    AIRSPY.converter = NULL;
+    AIRSPY.converter_state = NULL;
+}
+
+static bool parse_uintt(char *s, uint64_t *const value, int utype)
+{
+    uint_fast8_t base = 10;
+    char *s_end;
+    uint64_t u64_val;
+
+    if (s[0] == '-')  // likely forgot a parameter
+        return false;
+
+    if (!isdigit(s[0]))  // we expect at least one number
+        return false;
+
+    if( strlen(s) > 2 ) {  // check for hexadecimal
+        if(s[0] == '0') {
+            if( tolower(s[1]) == 'x' ) {
+                base = 16;
+                s += 2;
+            }
+        }
+    }
+
+    s_end = s;
+
+    u64_val = strtoull(s, &s_end, base);
+    if( (s != s_end) && (*s_end == 0) ) {
+        switch(utype) {
```

Comment:
I mean if the user gives you a value 257 for a uint8_t type then you will stuff 257 into 8 bits and get a result of 1 and not notice that it's out of range. You should check that the `unsigned long long` value is not too large for the target type.